File: android/src/main/java/expo/modules/demoexpo/DemoExpoModuleSharedObject.kt
package expo.modules.demoexpo

import expo.modules.kotlin.AppContext
import expo.modules.kotlin.sharedobjects.SharedObject

class DemoExpoModuleSharedObject(appContext: AppContext) : SharedObject(appContext) {
  var count: Int = 0

  override fun sharedObjectDidRelease() {
    super.sharedObjectDidRelease()
  }
}

File: src/DemoExpoModuleSharedObject.ts
import { SharedObject, useReleasingSharedObject } from 'expo-modules-core';

import DemoExpoModule from './DemoExpoModule';

export declare class DemoExpoModuleSharedObject extends SharedObject {
  count: number;
}

/**
 * Creates a new DemoExpoModuleSharedObject instance.
 * You are responsible for releasing it from memory by calling `release()` when done.
 */
export function createDemoExpoModuleSharedObject(): DemoExpoModuleSharedObject {
  return new DemoExpoModule.DemoExpoModuleSharedObject();
}

/**
 * A hook that creates a DemoExpoModuleSharedObject instance and automatically
 * releases it when the component unmounts.
 */
export function useDemoExpoModuleSharedObject(): DemoExpoModuleSharedObject {
  return useReleasingSharedObject(() => new DemoExpoModule.DemoExpoModuleSharedObject(), []);
}
